framework/test/gtest/ServiceHooksTest.cpp

```diff
@@ -237,20 +261,22 @@ namespace
             for (std::vector<ListenerInfo>::const_iterator iter = listeners.begin(); iter != listeners.end(); ++iter)
             {
                 listenerInfos.erase(*iter);
-                ordering.push_back(id * 10);
+                ordering_->push_back(id * 10);
             }
             lastRemoved = listeners.back();
         }
 
-        static std::vector<int> ordering;
+        std::vector<int>
```

Comment:
return a `std::shared_ptr<std::vector<int>>>` instead. this breaks the purpose of having a `std::shared_ptr<std::vector<int>>>`

Reply:
I'm happy to make that change. In this case, the ordering is a `std::shared_ptr<std::vector<int>>` owned by the hooks. In the case of `GetOrdering()`, we are just comparing the value of the `std::vector` to some expected ordering. It has to be shared in order to be owned and edited by multiple hooks. 

I am curious why the calling code needs, or wants, to have a `shared_ptr<std::vector<int>>` to this ordering rather than just owning a `std::shared_ptr<ServiceHooksObject>` to the object that owns `shared_ptr<vector<int>> ordering_`. As long as the calling code has a `std::shared_ptr<ServiceHooksObject>`, that object and therefore the `ordering_` will not die

Reply:
I'm not sure what you mean by `calling code` in this case. Taking a step back, these tests are verifying the order of events delivered to multiple `ServiceEventListenerHook` and `ServiceFindHook` objects. Since `TestServiceFindHook` and `TestServiceEventListenerHook` derive from abstract base classes, could you use [gmock to check for a specific sequence of method calls](https://github.com/google/googletest/blob/main/docs/gmock_for_dummies.md#ordered-vs-unordered-calls-orderedcalls)?

Reply:
@jeffdiclemente While I agree that you could make mocks, there is also a lot of other things going on in the `Forward()`, `Find()` etc functions other than just verifying order of calls -- see [here](https://github.com/CppMicroServices/CppMicroServices/blob/e430df5df499098d56cf2aab3a1cee2d8be9dc7b/framework/test/gtest/ServiceHooksTest.cpp#L89C9-L90C10) for example. 

I read through the page on mocks and looked around and I think you are not allowed to program behavior for these mock functions. Given that, we would have to rearrange these tests significantly in some way (though I am not sure how off the top of my head) to verify what is currently verified.

Additionally we could verify that calls are made to `Find()`, `Added()`, `Removed()` etc. in the correct order, but the variable currently being placed into the vector is not an input but rather a private variable of the mock class. Given this, you don't have access to the `id` from the signature of the function call to use the syntax `EXPECT_CALL(mock, Removed(1));` and so the order would purely be which function is called, not which function with which input. 

Let me know if I am missing something here because ideally I agree, mocks would be the way forward.

Reply:
I agree, there is a lot of different behaviors being tested in the tests. I suggest pulling out the ordering related tests into separate test points and using mocks for those tests.

you can add custom behavior to your mocks. There are a few methods, based on specific needs detailed [here](http://google.github.io/googletest/gmock_cook_book.html#DelegatingToFake). Also, take a look at [Mocks.hpp](https://github.com/CppMicroServices/CppMicroServices/blob/e430df5df499098d56cf2aab3a1cee2d8be9dc7b/compendium/DeclarativeServices/test/gtest/Mocks.hpp#L330) for some examples of doing this.

As it relates to the variable being placed into the `ordering` vector, from what I can see the variable is used only as a means to check for the correct ordering. If using mocks allows to check for correct ordering of callback calls, is the variable placed in the `ordering` vector even needed? Is the `ordering` vector itself even needed?

Reply:
Darn, I missed this implementation ability in my looking around. Yes this seems like a good idea. The ordering variable is definitely not needed as long as we can verify that the correct instance of these functions is being called (which again is dependent on member variables, ie `id`, rather than arguments to the function call). I will look into a way to verify this behavior and go from there but separating out these tests feels like a good idea.

Reply:
You're right this should work no problem. I'll work on this before our meeting later this morning